openssh/libmatchers.jinja

```diff
@@ -0,0 +1,202 @@
+{#- -*- coding: utf-8 -*- #}
+{#- vim: ft=jinja #}
+
+{#- Get the `tplroot` from `tpldir` #}
+{%- set tplroot = tpldir.split("/")[0] %}
+{%- from tplroot ~ "/libsaltcli.jinja" import cli %}
+
+{%- set query_map = {
+      "C": "config.get",
+      "G": "grains.get",
+      "I": "pillar.get",
+    } %}
+
+{#- When no part before `@` is provided: #}
+{#-   - define a filename path, noted `F` #}
+{#-   - use `salt["config.get"]`, noted `C` #}
+{#-   - use colon `:` delimiter for querying #}
+{%- set _defaults = {
+      "type": "F",
+      "query_type": "C",
+      "query_delimiter": ":"
+    } %}
+
+{%- macro parse_matchers(
+      matchers,
+      config_get_strategy=None,
+      log_prefix="libmatchers: "
+    ) %}
+{#-   matcher format is `[<TYPE>[:<OPTION>[:DELIMITER]]@]<KEY>` #}
+{#-   each matcher has a type: #}
+{#-   - `F` to build a file name (the default when no type is set) #}
+{#-   - `C` to lookup values with `config.get` #}
+{#-   - `G` to lookup values with `grains.get` #}
+{#-   - `I` to lookup values with `pillar.get` #}
+{#-   The `FILE` type option can define query type to build the file name: #}
+{#-   - `C` for query with `config.get` (the default when to query type is set) #}
+{#-   - `G` for query with `grains.get` #}
+{#-   - `I` for query with `pillar.get` #}
+{#-   With `DELIMITER`, you can choose a different delimiter when doing queries #}
+{%-   set parsed_matchers = [] %}
+{%-   for matcher in matchers %}
+{%-     do salt["log.debug"](
+          log_prefix
+          ~ "process matcher: "
+          ~ "'" ~ matcher ~ "'"
+        ) %}
+
+{%-     set parsed = {} %}
+{%-     set matcher_parts = matcher.split('@') %}
+{%-     if matcher_parts | length == 1 %}
+{#-       By default we load YAML files for config looked up by `config.get` #}
+{%-       do parsed.update(
+            {
+              "type": _defaults["type"],
+              "option": None,
+              "query_method": query_map[_defaults["query_type"]],
+              "query_delimiter": _defaults["query_delimiter"],
+              "query": matcher
+            }
+          ) %}
+{%-       do salt["log.debug"](
+            log_prefix
+            ~ "use built-in defaults for matcher:\n"
+            ~ parsed | yaml(False) | indent(4, first=True)
```

Comment:
Running into the same error as mentioned in https://github.com/saltstack-formulas/openvpn-formula/pull/134#issuecomment-748871217.

Here are the failing jobs:

* [default-debian-9-3000-3-py3](https://gitlab.com/myii/openssh-formula/-/jobs/953110213)
* [default-ubuntu-1604-3000-3-py2](https://gitlab.com/myii/openssh-formula/-/jobs/953110215)

There are 6 occurrences of `first=True` in this file.

Reply:
@baby-gnu Skipped around that by removing all of the occurrences of `first=True` and now have encountered a different, more significant failure:

```sls
local:
    Data failed to compile:
----------
    Rendering SLS 'base:openssh._mapdata' failed: Jinja syntax error: no test named 'eq'
/tmp/kitchen/var/cache/salt/minion/files/base/openssh/libmapstack.jinja(211):
---
[...]
     %}

{%-         for yaml_name in yaml_names %}
{#-           Make sure to have a `.yaml` extension #}
{#-           Use `.rpatition` to strip last `.yaml` in `dir.yaml/file.yaml` #}
{%-           set yaml_filename = [    <======================
         yaml_dir.rstrip("/"),
         yaml_name.rpartition(".yaml")
         | reject("eq", ".yaml")
         | join
         ~ ".yaml"
[...]
---
----------
    Rendering SLS 'base:openssh.config' failed: Jinja syntax error: no test named 'eq'
/tmp/kitchen/var/cache/salt/minion/files/base/openssh/libmapstack.jinja(211):
---
[...]
     %}

{%-         for yaml_name in yaml_names %}
{#-           Make sure to have a `.yaml` extension #}
{#-           Use `.rpatition` to strip last `.yaml` in `dir.yaml/file.yaml` #}
{%-           set yaml_filename = [    <======================
         yaml_dir.rstrip("/"),
         yaml_name.rpartition(".yaml")
         | reject("eq", ".yaml")
         | join
         ~ ".yaml"
[...]
---
----------
    Rendering SLS 'base:openssh.known_hosts' failed: Jinja syntax error: no test named 'eq'
/tmp/kitchen/var/cache/salt/minion/files/base/openssh/libmapstack.jinja(211):
---
[...]
     %}

{%-         for yaml_name in yaml_names %}
{#-           Make sure to have a `.yaml` extension #}
{#-           Use `.rpatition` to strip last `.yaml` in `dir.yaml/file.yaml` #}
{%-           set yaml_filename = [    <======================
         yaml_dir.rstrip("/"),
         yaml_name.rpartition(".yaml")
         | reject("eq", ".yaml")
         | join
         ~ ".yaml"
[...]
---
```

Here are the failing jobs for this error:

* [default-debian-9-3000-3-py3](https://gitlab.com/myii/openssh-formula/-/jobs/953128440)
* [default-ubuntu-1604-3000-3-py2](https://gitlab.com/myii/openssh-formula/-/jobs/953128442)

Reply:
Thanks a lot, I forgot the `first=True` issue.

Reply:
> Thanks a lot, I forgot the `first=True` issue.

@baby-gnu No problem.  However, the [`no test named 'eq'`](https://github.com/saltstack-formulas/openssh-formula/pull/191#discussion_r554438777) issue may be more difficult to resolve.

Reply:
* The `first=True` issue is solved by using only the second positional argument `True`.
* The `no test named 'eq'` error is solved by using `equalto`.

<details>
<summary>All platforms, except `centos-6` which can't be created, are verified on my local test bed</summary>

```Instance                              Driver  Provisioner  Verifier  Transport  Last Action    Last Error
default-debian-10-master-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-2004-master-py3        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-1804-master-py3        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-centos-8-master-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-fedora-32-master-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-fedora-31-master-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-opensuse-leap-152-master-py3  Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-amazonlinux-2-master-py3      Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-debian-10-3000-3-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-debian-9-3000-3-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-1804-3000-3-py3        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-centos-8-3000-3-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-centos-7-3000-3-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-fedora-31-3000-3-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-opensuse-leap-152-3000-3-py3  Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-amazonlinux-2-3000-3-py3      Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-1804-3000-3-py2        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-1604-3000-3-py2        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-arch-base-latest-3000-3-py2   Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-debian-10-2019-2-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-debian-9-2019-2-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-1804-2019-2-py3        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-ubuntu-1604-2019-2-py3        Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-centos-8-2019-2-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-centos-7-2019-2-py3           Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-fedora-31-2019-2-py3          Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-opensuse-leap-152-2019-2-py3  Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-amazonlinux-2-2019-2-py3      Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-centos-6-2019-2-py2           Docker  SaltSolo     Inspec    Ssh        <Not Created>  Kitchen::ShellOut::ShellCommandFailed
default-amazonlinux-1-2019-2-py2      Docker  SaltSolo     Inspec    Ssh        Verified       <None>
default-arch-base-latest-2019-2-py2   Docker  SaltSolo     Inspec    Ssh        Verified       <None>
```
</details>